runtime/jvmti/jvmtiHelpers.c

```diff
@@ -95,56 +95,108 @@ static UDATA watchedClassEqual (void *lhsEntry, void *rhsEntry, void *userData);
-	if (!J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) || ((targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject)) == NULL)) {
+#if JAVA_SPEC_VERSION >= 19
+	isVirtualThread = IS_VIRTUAL_THREAD(currentThread, threadObject);
+	if (isVirtualThread) {
+		if  (NULL != isVirtual) {
+			*isVirtual = TRUE;
+		}
+		/* Make sure the virtual thread stays alive while it is being used. */
+		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+		jint vthreadState = J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, threadObject);
+		j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(currentThread, threadObject);
+		if (NULL != carrierThread) {
+			targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, carrierThread);
+		}
+		isThreadAlive = (vthreadState != JVMTI_VTHREAD_STATE_NEW) && (vthreadState != JVMTI_VTHREAD_STATE_TERMINATED);
+	} else
+#endif /* JAVA_SPEC_VERSION >= 19 */
+	{
+		targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, threadObject);
+		isThreadAlive = J9VMJAVALANGTHREAD_STARTED(currentThread, threadObject) && (NULL != targetThread);
+	}
+
+	if (!isThreadAlive) {
 		if (mustBeAlive) {
+#if JAVA_SPEC_VERSION >= 19
+			if (isVirtualThread) {
+				omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
+			}
+#endif /* JAVA_SPEC_VERSION >= 19 */
 			omrthread_monitor_exit(vm->vmThreadListMutex);
 			return JVMTI_ERROR_THREAD_NOT_ALIVE;
 		}
 	}
 
 	*vmThreadPtr = targetThread;
-	if (targetThread != NULL) {
-		++(targetThread->inspectorCount);
+	if (NULL != targetThread) {
+		targetThread->inspectorCount += 1;
 	}
+#if JAVA_SPEC_VERSION >= 19
+	if (isVirtualThread) {
+		U_64 vthreadInspectorCount = J9OBJECT_U64_LOAD(currentThread, threadObject, vm->virtualThreadInspectorCountOffset);
+		J9OBJECT_U64_STORE(currentThread, threadObject, vm->virtualThreadInspectorCountOffset, (vthreadInspectorCount + 1));
+		omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
+	}
+#endif /* JAVA_SPEC_VERSION >= 19 */
 	omrthread_monitor_exit(vm->vmThreadListMutex);
 	return JVMTI_ERROR_NONE;
 }
 
 
 
 void
-releaseVMThread(J9VMThread * currentThread, J9VMThread * targetThread)
+releaseVMThread(J9VMThread *currentThread, J9VMThread *targetThread, jthread thread, BOOLEAN isVirtual)
 {
-	if ((currentThread != targetThread) && (targetThread != NULL)) {
-		J9JavaVM * vm = targetThread->javaVM;
-
-		/* Release this thread (allow it to die) now that we are no longer inspecting it */
+#if JAVA_SPEC_VERSION >= 19
+	if (isVirtual && (NULL != thread)) {
+		j9object_t threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+		if (NULL != threadObject) {
+			J9JavaVM *vm = targetThread->javaVM;
+			U_64 vthreadInspectorCount = 0;
+			/* Release the virtual thread (allow it to die) now that we are no longer inspecting it. */
+			omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+			vthreadInspectorCount = J9OBJECT_U64_LOAD(currentThread, threadObject, vm->virtualThreadInspectorCountOffset) - 1;
```

Comment:
I suggest to add an assertion that the old inspector count is not 0 before it is decremented here. And add an assertion that the new inspector count is not 0 after incremented in `getVMThread()`.

Reply:
Updated.